### PR TITLE
Add basic Yarn 2 support

### PR DESCRIPTION
**Pull request by dchambers**
As [described here](https://github.com/yarnpkg/berry/issues/882), Yarn 2 doesn't support non-JavaScript binaries. A first improvement for `catch-uncomitted`, which I believe would give us everything but Zero Installs, is to just invoke the shell script from Node.js, which this PR does.

Alternatively, to be able to support Zero Installs as well, I could port the shell script to JavaScript if you are open to the idea? Something along these lines so we don't have to introduce any external dependencies maybe?:

```js
const { execSync } = require('child_process')

const defaultOptions = {
  catchNoGit: false,
  ignoreEol: '',
  extraArgs: [],
}

const nodeVersionbotFiles = [
  'package.json',
  'package-lock.json',
  'npm-shrinkwrap.json',
  'CHANGELOG.md',
  '.versionbot/CHANGELOG.yml',
]

try {
  const userOptions = process.argv.slice(2).reduce((options, arg) => {
    switch (arg) {
      case '--catch-no-git': {
        return {
          ...options,
          catchNoGit: true,
        }
      }

      case '--skip-node-versionbot-changes': {
        return {
          ...options,
          extraArgs: [
            ...options.extraArgs,
            nodeVersionbotFiles.map((path) => `:(exclude)${path}`),
          ],
        }
      }

      default: {
        if (arg.startsWith('exclude=')) {
          return {
            ...options,
            extraArgs: [
              ...options.extraArgs,
              arg.split('=')[1], // TODO: implement correctly
            ],
          }
        }

        throw new Error(
          'usage: catch-uncommitted [--catch-no-git] [--skip-node-versionbot-changes] [--exclude=<path>]'
        )
      }
    }
  }, defaultOptions)

  if (
    userOptions.catchNoGit &&
    execSync('hash git', { stdio: 'ignore' }).length !== 0
  ) {
    throw new Error('git not found, not running catch-uncommitted check')
  }

  // TODO: continue porting along these lines...
} catch (e) {
  console.error(e.message)
  process.exit(1)
}
```

### CODE REVIEW

**Comment by thgreasi** [comment]
Hey @dchambers,
Thanks for your contribution to this module.
In order to get our CI tests to pass we follow some commit guidelines, and that's probably the reason that they fail atm.
It would be great if you could amend your current commit and add a `Change-type: major` in it's body.
The reasoning on marking this as a major is just b/c this also drops support for node < v8 (someone could argue that's EOL though, but it's a good practice to avoid surprises on consumers).

In short you can run `git commit --amend` and in the editor change the text to something like:

```
Add basic Yarn 2 support 

Change-type: major
```
and force push.

Let me know if you would prefer me taking care of that for you.

File: catch-uncomitted.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env node
+const { execFileSync } = require('child_process')
+const path = require('path')
+
+const argv = process.argv.slice(2)
+
+try {
+  execFileSync(`${path.join(__dirname, 'catch-uncomitted.sh')}`, argv, {
```
Comment (thgreasi) [inline catch-uncomitted.js]: I think that you don't actually need the extra string template here, since `path.join` already returns a string, right?
```suggestion
  execFileSync(path.join(__dirname, 'catch-uncomitted.sh'), argv, {
```

File: catch-uncomitted.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    stdio: 'inherit',
+  })
+} catch (error) {
+  process.exit(error.status)
```
Comment (pdcastro) [inline catch-uncomitted.js]:
```suggestion
  process.exitCode = error.status
```
 
Unless there's a specific reason to call `process.exit()`, I understand that the "graceful exit" approach is to set `process.exitCode` instead. See docs: https://nodejs.org/docs/latest-v8.x/api/process.html#process_process_exit_code

File: catch-uncomitted.js (hunk 1 above)
Comment (pdcastro) [reply catch-uncomitted.js]: There is also a typo in the script name: `uncomitted` with a single `m` instead of `uncommitted` with double `m` -- which suggests that this PR has not been tested...

File: package.json [hunk 3]
```diff
@@ -3,7 +3,10 @@
...
   "description": "Script to catch uncommitted/unversioned files, for CI",
   "bin": {
-    "catch-uncommitted": "catch-uncommitted.sh"
+    "catch-uncommitted": "catch-uncommitted.js"
+  },
+  "engines": {
+    "node": ">=8.0.0"
   },
   "scripts": {
     "test": "./catch-uncommitted.sh --skip-node-versionbot-changes"
```
Comment (pdcastro) [inline package.json]:
```suggestion
    "test": "node ./catch-uncommitted.js --skip-node-versionbot-changes"
```
Maybe.

File: catch-uncomitted.js (hunk 1 above)
Comment (dchambers) [reply catch-uncomitted.js]: True! I think as you identify below the problem is that the test isn't actually using the new script. The script was only tested as working within our monorepo, where the spelling mistake originated from.

Comment (dchambers) [reply catch-uncomitted.js:8]: Good shout!

File: package.json (hunk 3 above)
Comment (dchambers) [reply package.json]: Good shout! I've amended slightly so as not to explicitly use `node` though, such that that test also verifies the `#!/usr/bin/env node`, and that the correct file execution permissions have been set.

**Comment by dchambers** [comment]
Thanks for the excellent feedback here, and I'm sorry I left this one languishing for so long on my side. All checks now green though, as predicted.

**Comment by dchambers** [comment]
`yarn test` run with no changes made:

image

`yarn test` run after a single change has been made:

image

File: package.json (hunk 3 above)
Comment (pdcastro) [reply package.json]: This is fine, as `catch-uncommitted` was already assuming a Unix environment by relying on a bash script. Windows cmd.exe / PowerShell (not WSL / Cygwin / MSYS...) does not understand [shebangs](https://en.wikipedia.org/wiki/Shebang_(Unix)) (at least the last time I checked) or file execution permissions, and would require `node ./catch-uncommitted.js`. Which brings me to the following point:

> Alternatively, to be able to support Zero Installs as well, I could port the shell script to JavaScript if you are open to the idea? Something along these lines so we don't have to introduce any external dependencies maybe?: [...]

Personally I didn't even know what [Zero Installs](https://yarnpkg.com/features/zero-installs) was about (I just googled it), :-) but I am always in favour of replacing shell script code with code in any semi-decent programming language, in this case JavaScript. And, by doing so, this PR would also be adding support for Windows (as long as the test script started with `node ...`). :-)  So yes, if you are still up to it, I would encourage replacing the bash script with a JS script like you suggested. A note regarding this line:

```
    execSync('hash git', { stdio: 'ignore' }).length !== 0
```

`hash git` is a bash/shell thing, but I understand that the only purpose of that line is to find out if `git` is installed / available in the PATH env var. So I think it could be replaced with `execSync('git', ...)` -- basically we would want `execSync` to fail if it cannot execute `git`.

Regarding:
```
  "engines": {
    "node": ">=8.0.0"
```

This PR being semver-major, as far as I am aware / concerned we could require `"node": ">=10.0.0"`, which would allow you to rely on ES2018 features and extended availability of the promises interface, e.g. `let { promises: fs } = require('fs')` (if you wanted).

If you prefer to just get this PR merged as it is (if you no longer have time to do the JS porting), that's fine too, 👍  just let us know.

**Comment by dchambers** [comment]
Hi @thgreasi / @pdcastro, just a reminder that this is ready once you have a spare moment.

**Comment by pdcastro** [comment]
> just a reminder that this is ready once you have a spare moment.

@dchambers, did you notice https://github.com/balena-io-modules/catch-uncommitted/pull/16#discussion_r623832263 ?

It ends with this question:
> If you prefer to just get this PR merged as it is (if you no longer have time to do the JS porting), that's fine too, 👍    just let us know.

Let us know. :-)

**Comment by dchambers** [comment]
Hi @thgreasi / @pdcastro, just a reminder that this is ready once you have a spare moment.

> > just a reminder that this is ready once you have a spare moment.
> 
> @dchambers, did you notice [#16 (comment)](https://github.com/balena-io-modules/catch-uncommitted/pull/16#discussion_r623832263) ?
> 
> It ends with this question:
> 
> > If you prefer to just get this PR merged as it is (if you no longer have time to do the JS porting), that's fine too, +1    just let us know.
> 
> Let us know. :-)

Eeek, I'm getting too old for this! I've had this tab open all week, refreshing daily to see if there was any progress, and only now do I see the nested comment above! :roll_eyes: 

I'm still wanting to finish off the port to JavaScript I started in the comment above, but I probably won't need it for another month as the upgrade to Yarn 2 is already a large move for us, so will want some bedding in before we take the next step. However, updating to Node 10 now (as part of this PR) is a cracking idea :+1:, since it saves another major version change down the line.

I'll update the `engine` config now...

**Comment by dchambers** [comment]
@pdcastro, I've updated the `engine` config to give us Node 10 support, and re-run the test locally, which is all still working:

image

**Review by pdcastro — APPROVED** [review]
